warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuildingVisitor.java

```diff
@@ -238,24 +244,47 @@ protected JexlNode flattenTree(JexlNode currentNode, JexlNode newNode, Object da
             JexlNode dereferenced = JexlASTHelper.dereference(node);
             
             if (acceptableNodesToCombine(currentNode, dereferenced, !node.equals(dereferenced))) {
-                flattenTree(dereferenced, newNode, data);
+                flattenTree(dereferenced, newNode);
             } else {
                 newNode.jjtAddChild(node, newNode.jjtGetNumChildren());
                 node.jjtSetParent(newNode);
             }
         }
+    }
+    
+    protected void flattenTree(JexlNode currentNode, JexlNode newNode) {
+        if (!currentNode.getClass().equals(newNode.getClass())) {
+            log.warn("newNode is not the same type as currentNode ... something has probably gone horribly wrong");
+        }
         
-        return newNode;
+        LinkedList<JexlNode> stack1 = new LinkedList<>();
+        LinkedList<JexlNode> stack2 = new LinkedList<>();
+        stack1.push(currentNode);
+        do {
+            JexlNode node = stack1.pop();
+            for (int i = node.jjtGetNumChildren() - 1, j = 0; i >= 0; i--) {
+                JexlNode child = node.jjtGetChild(i);
+                JexlNode dereferencedChild = JexlASTHelper.dereference(child);
+                
+                if (acceptableNodesToCombine(node, dereferencedChild, !child.equals(dereferencedChild))) {
+                    stack1.add(j++, dereferencedChild);
+                } else {
+                    stack2.push(child);
+                }
+            }
+        } while (!stack1.isEmpty());
+        do {
+            JexlNode node = (JexlNode) stack2.pop().jjtAccept(this, null);
+            newNode.jjtAddChild(node, newNode.jjtGetNumChildren());
+            node.jjtSetParent(newNode);
+        } while (!stack2.isEmpty());
     }
     
-    protected boolean acceptableNodesToCombine(JexlNode currentNode, JexlNode newNode, boolean isWrapped) {
+    protected static boolean acceptableNodesToCombine(JexlNode currentNode, JexlNode newNode, boolean isWrapped) {
```

Comment:
Does this need to be static?  Making it static makes the class less extensible.

Reply:
reverted